src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -651,10 +654,48 @@ private static string GenerateClassName(TypeDeclarationSyntax typeDeclaration)
                                 return (null, true);
                             }
                         }
+
+                        shadowedNames.Add(fs.Name);
                     }
 
                     onMostDerivedType = false;
-                    classType = classType.BaseType;
+                    currentClassType = currentClassType.BaseType;
+                }
+
+                // We prioritize fields over primary constructor parameters and avoid warnings if both exist.
+                if (loggerField is not null)
+                {
+                    return (loggerField, false);
+                }
+
+                IEnumerable<IMethodSymbol> primaryConstructors = classType.InstanceConstructors
+                    .Where(ic => ic.DeclaringSyntaxReferences
+                        .Any(ds => ds.GetSyntax() is ClassDeclarationSyntax));
+
+                foreach (IMethodSymbol primaryConstructor in primaryConstructors)
+                {
+                    foreach (IParameterSymbol parameter in primaryConstructor.Parameters)
+                    {
+                        if (IsBaseOrIdentity(parameter.Type, loggerSymbol))
+                        {
+                            if (shadowedNames.Contains(parameter.Name))
```

Comment:
>shadowedNames [](http://example.com/codeflow?start=32&length=13)

I am not sure if shadowedNames is needed at all? If we encountered any field of type `loggerSymbol`, we'll not reach here at all because we'll return `return (loggerField, false);`.

Reply:
> If we encountered any field of type `loggerSymbol`, we'll not reach here at all because we'll return (loggerField, false);.

Yes, that is correct. This is for the case where we do not encounter any fields of type `loggerSymbol`, but where one of the non-logger fields shadows a logger primary constructor parameter. Here's a relevant test case:

```csharp
partial class C(ILogger logger)
{
    private readonly object logger = logger;

    [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = ""M1"")]
    public partial void M1();
}
```

In this case, we can't use the `logger` primary constructor parameter, since there's no way to refer to it.